#  <CENTER>INTRODUCCIÓN A PYTHON - CEFIP<CENTER>
   
## <CENTER>OCTUBRE 2023<CENTER>

# <CENTER> Repaso clase anterior

- Anaconda, Jupyter Notebook, GitHub.
- Tipos de datos
- Operaciones con datos
- Variables
- Funciones
- Condicionales
- Módulos


# <CENTER> BUCLES

Existen dos tipos de bucles comunmente usados en python, los `for` loops y los `while` loops. 

## FOR LOOPS

Con los `for` loops se desea iterar sobre una secuencia (como una lista, una tupla, una cadena de texto o un rango numérico). Serán muy similares a los loops que solemos usar en Stata (`foreach`, `forvalues`).

In [2]:
numeros=[7,5,6,15,20]
y=0
for i in numeros: 
    y+=i
    print(f"La suma de los números es {y} hasta {i}")    

La suma de los números es 7 hasta 7
La suma de los números es 12 hasta 5
La suma de los números es 18 hasta 6
La suma de los números es 33 hasta 15
La suma de los números es 53 hasta 20


In [3]:
numeros=[7,5,6,15,20]
y=0
for i in numeros: 
    y+=i
    print(f"{i}")   

7
5
6
15
20


**Notar:** en el bloque anterior iteramos sobre una **lista** cuyos valores estaban predeterminados. Cualquier tipo de lista puede ser usada para un ciclo `for`.

**Notar 2:** poniendo la f adentro del print pudimos insertar valores de variables u expresiones directamente en una cadena de texto sin tener que concatenar manualmente las partes de la cadena.

In [4]:
for x in range(0,16,2): 
    print(x)
print("Estoy fuera del bucle")

0
2
4
6
8
10
12
14
Estoy fuera del bucle


**Recordar**: range(desde, hasta, paso) no incluye el `hasta`. 

Como se ve, puedo iterar usando otras letras, no solo i.


In [5]:
veces = int(input("¿Cuántas veces quiere que le salude? "))
for i in range(veces):
    print("Hola ", end="")
print()
print("Adiós")

¿Cuántas veces quiere que le salude? 3
Hola Hola Hola 
Adiós


**Notar:** la función `input()` permite que tu programa interactúe con el usuario, solicitando información que el usuario debe proporcionar a través del teclado.

**Notar 2:** `input()` siempre devuelve una cadena de caracteres (string), al usar `int(input())`,  estamos convirtiendo esa cadena en un número entero. 

In [6]:
#Nos puede interesar ver la posición además del número sobre el que itera:
for indice, x in enumerate(range(-2,2), start=1):
    print(indice, x)

1 -2
2 -1
3 0
4 1


El índice generado por enumerate() comienza en 0 por defecto.

Con `start` le indicamos un valor inicial  diferente. 

La función `enumerate()` es útil cuando necesitas tanto el valor como el índice de los elementos en una secuencia durante la iteración.

In [7]:
tabla_2 = [x*2 for x in range(1,11)]

print(tabla_2)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


El ejemplo anterior nos muestra que podemos crear una lista por ejemplo a partir de un bucle. Con esta sintaxis también podemos crear una tupla, diccionario, etc.

## WHILE LOOPS

Los utilizaremos cuando el número de iteraciones no es conocido al comenzar. Si bien es Stata también existen, no los usamos tan seguido. 

Para comenzar aprovechemos para importar un módulo que puede ser de utilidad `random`. En este caso lo usaremos para partir de un número aleatorio entero. 

In [9]:
import random

In [14]:
n = 0
favorite = int(input("¿Cuál es su número preferido? "))
while n < 100:
    n += 1
    draw = random.randint(1, 49) # e.g. German lottery
    if draw == favorite:
        print("Salió mi número preferido :)")
        break
else:
    print("Mi número preferido no apareció:(")
print(f"¡Probé {n} veces!")


¿Cuál es su número preferido? 12
Salió mi número preferido :)
¡Probé 74 veces!


**Break**: Cuando se encuentra una declaración `break` dentro de un bucle, el programa sale  del bucle y continúa ejecutando el código que sigue después del bucle.

# <CENTER>TRABAJAR CON DATOS EN PYTHON
    
# <CENTER> PANDAS
    
pandas es una biblioteca esencial para la manipulación y análisis de datos en Python debido a su facilidad de uso, flexibilidad y potencia. Se integra con otras herramientas de ciencia de datos y machine learning.

In [17]:
!pip install xlwt
!pip install pandas

Lo primero que tenemos que hacer con una librería es instalarla en caso de que no esté (Pandas es una de las librerías que se encuentra preinstalada con Anaconda). No es necesario volverlo a instalar si ya lo tenemos instalado. 

In [19]:
import pandas as pd
import numpy as np

In [20]:
dir(pd)

['ArrowDtype',
 'BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt8Dtype',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_is_numpy_dev',
 '_libs',
 '_testing',
 '_typing',
 '_version',
 'annotations',
 'api',
 'array',
 'arrays',
 'bdate_range',
 'compat',
 'concat',
 'c

Una vez instalado debemos importar la biblioteca, esto hay que hacerlo  **todas las veces que abrimos el notebook**.

**pd** es la abreviatura (o alias) tradicional para pandas en la literatura. 

# Estructuras de datos de Pandas:
## Series

- Será una columna unidimensional, donde podemos tener todo tipo de datos.

- Se suele mostrar con los índices a la izquierda, que por default van de 0 a N-1. 

In [25]:
#Comencemos por crear una lista en Pandas

proyectos_CEFIP=pd.Series(["impuestos provinciales", "impuestos municipales", "tarifas de servicios", "gasto por jurisdiccion"])
proyectos_CEFIP


0    impuestos provinciales
1     impuestos municipales
2      tarifas de servicios
3    gasto por jurisdiccion
dtype: object

La sintaxis de creación de series sigue la estructura: 

`s = pd.Series(datos, index=index)`

En el caso anterior, no asignamos nada al `index`, y le asignó el default.



In [26]:
proyectos_CEFIP2=pd.Series(["impuestos provinciales", "impuestos municipales", "tarifas de servicios"], index=["a", "b", "a"])
proyectos_CEFIP2

a    impuestos provinciales
b     impuestos municipales
a      tarifas de servicios
dtype: object

In [27]:
unidades_geograficas=pd.Series(["5 provincias argentinas", "135 municipios bonaerenses", "Empresas distribuidoras"])
unidades_geograficas

0       5 provincias argentinas
1    135 municipios bonaerenses
2       Empresas distribuidoras
dtype: object

IG: matchea por índice, necesariamente deben tener el mismo formato (integer, string, etc)

## DataFrames

Los **DataFrames** son una estructura de datos tabular bidimensional con etiquetas en las línea y columnas. Son la estructura más común en Pandas.


Como las Series, los DataFrames son capaces de almacenar cualquier tipo de datos y cada columna puede tener un tipo distinto. La forma básica de
creación de un DataFrame es:

`df = pd.DataFrame(datos, index=index, columns=columns)`

El argumento datos puede ser por ejemplo un diccionario, una lista, una Series u otro DataFrame.

In [30]:
# Creando un dataframe de autos y colores.
data_proyectos = pd.DataFrame({"Proyecto": proyectos_CEFIP, 
                               "Unidades geográficas": unidades_geograficas})

In [31]:
data_proyectos

Proyecto        Unidades geográficas
0  impuestos provinciales     5 provincias argentinas
1   impuestos municipales  135 municipios bonaerenses
2    tarifas de servicios     Empresas distribuidoras
3  gasto por jurisdiccion                         NaN

### Los diccionarios son parecidos a las listas pero cada elemento implica clave-valor

Notar que cuando los índices no coinciden (sea el default como en este caso o en otro), es decir que no puede matchear las unidades, encontramos un valor missing, que en Pandas la identificaremos con `NaN`

In [32]:
data_municipios = pd.DataFrame({"Proyecto": ["Adolfo Alsina","Alberti", "Arrecifes", "Avellaneda"],
                                "Poblacion": [17460, 12650, 32229, 369556],
                               "Act. Económica": ["Agricultura", "Agricultura", "Agricultura", "Manufactura"]})
data_municipios

Proyecto  Poblacion Act. Económica
0  Adolfo Alsina      17460    Agricultura
1        Alberti      12650    Agricultura
2      Arrecifes      32229    Agricultura
3     Avellaneda     369556    Manufactura

En el ejemplo anteriopr creamos el DataFrame a partir de un diccionario de listas. Se asingó el índice por default (como pasaba con la serie). 

## Importar datos

Importar datos es una de las formas de crear un dataframe, a partir de un archivo ya existente (en formato csv, xls, dta, etc.). 

In [33]:
dataset=pd.read_csv(r"C:\Users\isidr\Dropbox\GitHub\CEFIP\Clase 2\db.csv", sep=";")

In [34]:
dataset


Nombre                  Motor   Año  Kilometraje  Cero_km  \
0            Jetta Variant        Motor 4.0 Turbo  2003      44410.0    False   
1                   Passat           Motor Diesel  1991       5712.0    False   
2                 Crossfox        Motor Diesel V8  1990      37123.0    False   
3                      DS5        Motor 2.4 Turbo  2020          NaN     True   
4         Aston Martin DB4        Motor 2.4 Turbo  2006      25757.0    False   
..                     ...                    ...   ...          ...      ...   
253           Phantom 2013               Motor V8  2014      27505.0    False   
254  Cadillac Ciel concept               Motor V8  1991      29981.0    False   
255              Clase GLK  Motor 5.0 V8 Bi-Turbo  2002      52637.0    False   
256       Aston Martin DB5           Motor Diesel  1996       7685.0    False   
257                  Macan        Motor Diesel V6  1992      50188.0    False   

                                            Accesorios     Valor  
0    ['Llantas de aleación', 'Cerraduras eléctricas...  17615.73  
1    ['Pantalla multimedia', 'Techo panorámico', 'F...  21232.39  
2    ['Piloto automático', 'Control de estabilidad'...  14566.43  
3    ['Cerraduras eléctricas', '4 X 4', 'Vidrios el...  24909.81  
4    ['Llantas de aleación', '4 X 4', 'Pantalla mul...  18522.42  
..                                                 ...       ...  
253  ['Control de estabilidad', 'Piloto automático'...  10351.92  
254  ['Bancos de cuero', 'Tablero digital', 'Sensor...  10333.41  
255  ['Llantas de aleación', 'Control de tracción',...  13786.81  
256  ['Aire acondicionado', '4 X 4', 'Cambio automá...  24422.18  
257  ['Pantalla multimedia', 'Techo panorámico', 'V...  18076.29  

[258 rows x 7 columns]

In [36]:
## Analizando lo que tenemos/ lo que nos muestra: 
- Primera columna con los índices. 
- Nos muestra las primeras 5 y últimas 5 observaciones.
- Cant de filas (obs), cant de columnas (variables). 
- Punto es separador decimal.
- Tenemos distintos tipos en todas las columnas.

SyntaxError: invalid syntax (3299803185.py, line 2)

In [39]:
df = pd.read_stata(r"C:\Users\isidr\Dropbox\GitHub\CEFIP\Clase 2\datos_bp.dta") 
df

patient     sex agegrp    when   bp
0          1    Male  30-45  Before  143
1          1    Male  30-45   After  153
2          2    Male  30-45  Before  163
3          2    Male  30-45   After  170
4          3    Male  30-45  Before  153
..       ...     ...    ...     ...  ...
235      118  Female    60+   After  174
236      119  Female    60+  Before  149
237      119  Female    60+   After  151
238      120  Female    60+  Before  185
239      120  Female    60+   After  163

[240 rows x 5 columns]

In [40]:
dataset

Nombre                  Motor   Año  Kilometraje  Cero_km  \
0            Jetta Variant        Motor 4.0 Turbo  2003      44410.0    False   
1                   Passat           Motor Diesel  1991       5712.0    False   
2                 Crossfox        Motor Diesel V8  1990      37123.0    False   
3                      DS5        Motor 2.4 Turbo  2020          NaN     True   
4         Aston Martin DB4        Motor 2.4 Turbo  2006      25757.0    False   
..                     ...                    ...   ...          ...      ...   
253           Phantom 2013               Motor V8  2014      27505.0    False   
254  Cadillac Ciel concept               Motor V8  1991      29981.0    False   
255              Clase GLK  Motor 5.0 V8 Bi-Turbo  2002      52637.0    False   
256       Aston Martin DB5           Motor Diesel  1996       7685.0    False   
257                  Macan        Motor Diesel V6  1992      50188.0    False   

                                            Accesorios     Valor  
0    ['Llantas de aleación', 'Cerraduras eléctricas...  17615.73  
1    ['Pantalla multimedia', 'Techo panorámico', 'F...  21232.39  
2    ['Piloto automático', 'Control de estabilidad'...  14566.43  
3    ['Cerraduras eléctricas', '4 X 4', 'Vidrios el...  24909.81  
4    ['Llantas de aleación', '4 X 4', 'Pantalla mul...  18522.42  
..                                                 ...       ...  
253  ['Control de estabilidad', 'Piloto automático'...  10351.92  
254  ['Bancos de cuero', 'Tablero digital', 'Sensor...  10333.41  
255  ['Llantas de aleación', 'Control de tracción',...  13786.81  
256  ['Aire acondicionado', '4 X 4', 'Cambio automá...  24422.18  
257  ['Pantalla multimedia', 'Techo panorámico', 'V...  18076.29  

[258 rows x 7 columns]

## Repasando:

- Importar de csv `data=pd.read_csv('archivo.csv')`
- Importar de excel `data = pd.read_excel('archivo.xlsx', sheet_name='hoja1')`
- Importar de stata `data = pd.read_stata('archivo.dta')`
- Importar de una web `data = pd.read_html('link')`


## Tener en cuenta: 
En **windows** los directorios son cadena de texto con barras invertidas (\) como separadores. Las barras invertidas son caracteres de escape en las cadenas de Python, por lo que es necesario utilizar el prefijo "r" antes de la cadena para indicar que es una "cadena cruda" y que los caracteres de escape no deben interpretarse. (En **Mac** no se tiene este problema). 


### Disgresión: caracteres de escape 
Dentro de las cadenas de caracteres `\n` implica un salto de línea, `\t` una tabulación, y si queremos escribir una barra invertida en una cadena hay que poner `\\`

In [41]:
print("Hola \nMundo")
print("Hola \tMundo")
print("Hola \\Mundo")

Hola 
Mundo
Hola 	Mundo
Hola \Mundo


## Volviendo...

### Importemos desde una web

¿Qué pasa si me quiero importar una tabla desde wikipedia?

In [42]:
tablas = pd.read_html('https://es.wikipedia.org/wiki/Anexo:Partidos_de_la_provincia_de_Buenos_Aires', encoding='utf-8')
tablas

[    0                                                  1
 0 NaN  Este artículo o sección se encuentra desactual...,
      Bandera Escudo                 Partido Sección electoral  \
 0        NaN    NaN           Adolfo Alsina               6.ª   
 1        NaN    NaN  Adolfo Gonzales Chaves               6.ª   
 2        NaN    NaN                 Alberti               4.ª   
 3        NaN    NaN         Almirante Brown               3.ª   
 4        NaN    NaN               Arrecifes               2.ª   
 ..       ...    ...                     ...               ...   
 131      NaN    NaN           Vicente López               1.ª   
 132      NaN    NaN            Villa Gesell               5.ª   
 133      NaN    NaN               Villarino               6.ª   
 134      NaN    NaN                  Zárate               2.ª   
 135      NaN    NaN                   TOTAL               NaN   
 
     Códigos INDEC: divisiones político-administrativas/ gobiernos locales[2]​[3]​  \
 0 

In [44]:
tabla=tablas[1]
tabla

Bandera Escudo                 Partido Sección electoral  \
0        NaN    NaN           Adolfo Alsina               6.ª   
1        NaN    NaN  Adolfo Gonzales Chaves               6.ª   
2        NaN    NaN                 Alberti               4.ª   
3        NaN    NaN         Almirante Brown               3.ª   
4        NaN    NaN               Arrecifes               2.ª   
..       ...    ...                     ...               ...   
131      NaN    NaN           Vicente López               1.ª   
132      NaN    NaN            Villa Gesell               5.ª   
133      NaN    NaN               Villarino               6.ª   
134      NaN    NaN                  Zárate               2.ª   
135      NaN    NaN                   TOTAL               NaN   

    Códigos INDEC: divisiones político-administrativas/ gobiernos locales[2]​[3]​  \
0                                        06007/ 060007                              
1                                        06014/ 060014                              
2                                        06021/ 060021                              
3                                        06028/ 060028                              
4                                        06077/ 060077                              
..                                                 ...                              
131                                      06861/ 060861                              
132                                      06868/ 060868                              
133                                      06875/ 060875                              
134                                      06882/ 060882                              
135                                                NaN                              

                   Cabecera  Año de creación[4]​  \
0                    Carhué               1886.0   
1    Adolfo Gonzales Chaves               1916.0   
2                   Alberti               1910.0   
3                   Adrogué               1873.0   
4                 Arrecifes               1784.0   
..                      ...                  ...   
131                  Olivos               1905.0   
132            Villa Gesell               1978.0   
133                 Médanos               1886.0   
134                  Zárate               1854.0   
135                     NaN                  NaN   

                                                Origen  \
0                              Por ley provincial 1827   
1    Por Ley provincial 3632 – Con tierras de los p...   
2    Por ley provincial 3237 - Con tierras de los p...   
3    Por ley provincial 856 - Con tierras de los pa...   
4    Al designarse el primer alcalde - Recategoriza...   
..                                                 ...   
131  Por ley provincial 2959 - Con tierras del part...   
132  Por decreto ley provincial 9024 - Con tierras ...   
133                            Por ley provincial 1827   
134                Por decreto del gobernador Obligado   
135                                                NaN   

                        Intendente/a (2019-2023) Población (2010)[5]​  \
0           Javier Andrés (Juntos por el Cambio)               17.072   
1    Eduardo Marcelo Santillán (Frente de Todos)               12.047   
2                  Germán Lago (Frente de Todos)               10.654   
3          Mariano Cascallares (Frente de Todos)              552.902   
4           Javier Olaeta (Juntos por el Cambio)               29.044   
..                                           ...                  ...   
131                   Soledad Martínez (PRO-JxC)              269.420   
132            Gustavo Barrera (Frente de Todos)               31.730   
133     Carlos Bevilacqua (Acción por Villarino)               31.014   
134       Osvaldo Raúl Cáffaro (Frente de Todos)              114.269   
135                                          NaN           15.625.084   

    

## Exportación de los datos

Supongamos que queremos exportar los datos de la tabla anterior a distintos formatos (stata, excel).


In [47]:
tabla.to_csv(r"C:\Users\isidr\Dropbox\GitHub\CEFIP\Clase 2\exported-municipios.csv")

In [49]:
tabla.to_excel(r"C:\Users\isidr\Dropbox\GitHub\CEFIP\Clase 2\exported-municipios.xlsx")

In [50]:
tabla.to_stata(r"C:\Users\isidr\Dropbox\GitHub\CEFIP\Clase 2\exported-municipios.dta", version=118)

C:\Users\isidr\AppData\Local\Temp\ipykernel_11856\2039773429.py:1: InvalidColumnName: 
Not all pandas column names were valid Stata variable names.
The following replacements have been made:

    Sección electoral   ->   Sección_electoral
    Códigos INDEC: divisiones político-administrativas/ gobiernos locales[2]​[3]​   ->   Códigos_INDEC__divisiones_políti
    Año de creación[4]​   ->   Año_de_creación_4_​
    Intendente/a (2019-2023)   ->   Intendente_a__2019_2023_
    Población (2010)[5]​   ->   Población__2010__5_​
    Superficie (km²)[6]​   ->   Superficie__km___6_​
    Densidad (hab./km²)   ->   Densidad__hab__km__
    Cant. de concejales[7]​[8]​   ->   Cant__de_concejales_7_​_8_​
    Consejeros escolares[9]​   ->   Consejeros_escolares_9_​

If this is not what you expect, please make sure you have Stata-compliant
column names in your DataFrame (strings only, max 32 characters, only
alphanumerics and underscores, no Stata reserved words)

  tabla.to_stata(r"C:\Users\isidr\Dropbo

### Notar:
En stata, si no especifico versión 118, por default python toma stata 10, que no soporta la codificación de nuestro archivo.

## Exploración de los datos

Comencemos por entender qué hay dentro de nuestro dataframe

In [51]:
df.describe()

patient          bp
count  240.000000  240.000000
mean    60.500000  153.904167
std     34.712206   13.083702
min      1.000000  125.000000
25%     30.750000  144.000000
50%     60.500000  152.000000
75%     90.250000  162.250000
max    120.000000  185.000000

In [52]:
df

patient     sex agegrp    when   bp
0          1    Male  30-45  Before  143
1          1    Male  30-45   After  153
2          2    Male  30-45  Before  163
3          2    Male  30-45   After  170
4          3    Male  30-45  Before  153
..       ...     ...    ...     ...  ...
235      118  Female    60+   After  174
236      119  Female    60+  Before  149
237      119  Female    60+   After  151
238      120  Female    60+  Before  185
239      120  Female    60+   After  163

[240 rows x 5 columns]

`.describe()` nos da una vista rápida con información estadística de las columnas **numéricas**.

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   patient  240 non-null    int16   
 1   sex      240 non-null    category
 2   agegrp   240 non-null    category
 3   when     240 non-null    category
 4   bp       240 non-null    int16   
dtypes: category(3), int16(2)
memory usage: 2.1 KB


`.info` nos da información acerca del dataframe, aclarando qué tipo de datos tenemos (notar que 3 de nuestras columnas son categóricas), si tiene valores missings, cuántas filas y columnas tenemos, etc. 

Cuando una serie de datos se almacena con el tipo de datos `category`, Pandas le asigna un número entero único a cada valor distinto en la serie y mantiene un mapeo entre los valores originales y sus correspondientes números enteros. Esto permite ahorrar memoria. 

In [54]:
# Mostrar las primeros 5 filas de mi dataframe
df.head()

patient   sex agegrp    when   bp
0        1  Male  30-45  Before  143
1        1  Male  30-45   After  153
2        2  Male  30-45  Before  163
3        2  Male  30-45   After  170
4        3  Male  30-45  Before  153

In [55]:
# Mostrar las primeros n filas de mi dataframe
df.head(7)

patient   sex agegrp    when   bp
0        1  Male  30-45  Before  143
1        1  Male  30-45   After  153
2        2  Male  30-45  Before  163
3        2  Male  30-45   After  170
4        3  Male  30-45  Before  153
5        3  Male  30-45   After  168
6        4  Male  30-45  Before  153

In [56]:
# Mostrar las ultimas n filas de mi dataframe
df.tail(7)

patient     sex agegrp    when   bp
233      117  Female    60+   After  152
234      118  Female    60+  Before  165
235      118  Female    60+   After  174
236      119  Female    60+  Before  149
237      119  Female    60+   After  151
238      120  Female    60+  Before  185
239      120  Female    60+   After  163

### Estadísticas descriptivas

Si quiero que encontrar una característica de una columna en particular debo hacer `df['mi_variable'].estadistico()`

In [57]:
df["bp"].mean()

153.90416666666667

Para guardar el valor (tipo local):

In [64]:
a=df["bp"].mean()
a

153.90416666666667

In [58]:
df["patient"].median()

60.5

In [59]:
df["bp"].quantile(q = [0.25, 0.5, 0.75])

0.25    144.00
0.50    152.00
0.75    162.25
Name: bp, dtype: float64

In [60]:
df["bp"].quantile()

152.0

In [61]:
df['bp'].std()

13.08370154675559

In [62]:
df['bp'].std()

13.08370154675559

In [67]:
df['bp'].std(), df['patient'].mean()

(13.08370154675559, 60.5)

Quizás nos interese agrupar por una variable y comparar entre grupos:

In [66]:
df.groupby(["sex"]).mean()

TypeError: 'Categorical' with dtype category does not support reduction 'mean'

## Modificando los datos

Hasta ahora sabemos crear datos, importarlos, exportarlos y aprendimos algo sobre describirlos. Pero probablemente tengamos interés en modificarlos u hacer alguna operación con ellos. 

Podemos eliminar duplicados

In [68]:
df.drop_duplicates()

patient     sex agegrp    when   bp
0          1    Male  30-45  Before  143
1          1    Male  30-45   After  153
2          2    Male  30-45  Before  163
3          2    Male  30-45   After  170
4          3    Male  30-45  Before  153
..       ...     ...    ...     ...  ...
235      118  Female    60+   After  174
236      119  Female    60+  Before  149
237      119  Female    60+   After  151
238      120  Female    60+  Before  185
239      120  Female    60+   After  163

[240 rows x 5 columns]

Podemos ordenar según alguna variable:

In [69]:
#Orden ascendente
df.sort_values('bp')

patient     sex agegrp    when   bp
19        10    Male  30-45   After  125
31        16    Male  30-45   After  126
161       81  Female  46-59   After  127
137       69  Female  30-45   After  131
17         9    Male  30-45   After  131
..       ...     ...    ...     ...  ...
67        34    Male  46-59   After  184
109       55    Male    60+   After  185
97        49    Male    60+   After  185
238      120  Female    60+  Before  185
76        39    Male  46-59  Before  185

[240 rows x 5 columns]

In [70]:
#Orden descendente
df.sort_values('bp', ascending=False)

patient     sex agegrp    when   bp
97        49    Male    60+   After  185
238      120  Female    60+  Before  185
109       55    Male    60+   After  185
76        39    Male  46-59  Before  185
67        34    Male  46-59   After  184
..       ...     ...    ...     ...  ...
137       69  Female  30-45   After  131
129       65  Female  30-45   After  131
161       81  Female  46-59   After  127
31        16    Male  30-45   After  126
19        10    Male  30-45   After  125

[240 rows x 5 columns]

En nuestra base de datos el id es el paciente que estamos viendo, puede ser buena idea establecer la variable `patient` como índice. (Notar que me desaparece de las columnas)

In [71]:
df = df.set_index('patient')
print(df)

            sex agegrp    when   bp
patient                            
1          Male  30-45  Before  143
1          Male  30-45   After  153
2          Male  30-45  Before  163
2          Male  30-45   After  170
3          Male  30-45  Before  153
...         ...    ...     ...  ...
118      Female    60+   After  174
119      Female    60+  Before  149
119      Female    60+   After  151
120      Female    60+  Before  185
120      Female    60+   After  163

[240 rows x 4 columns]


Lo anterior puede interesarnos por ejemplo para poder **encontrar** al paciente n, para el que tenemos más de una observación (en este caso before y after) y ambas tienen el mismo **índice**. Lo haremos con la función `loc`:

In [72]:
df.loc[1]

sex agegrp    when   bp
patient                          
1        Male  30-45  Before  143
1        Male  30-45   After  153

También nos puede interesar quedarnos con una **posición** en particular

In [77]:
df.iloc[0]

sex         Male
agegrp     30-45
when      Before
bp           143
Name: 1, dtype: object

Quizás nos interesa quedarnos con los índices hasta un valor:

In [78]:
df.loc[:3]

sex agegrp    when   bp
patient                          
1        Male  30-45  Before  143
1        Male  30-45   After  153
2        Male  30-45  Before  163
2        Male  30-45   After  170
3        Male  30-45  Before  153
3        Male  30-45   After  168

Notar que en el ejemplo anterior, me toma los índices **hasta el 3 inclusive**. En el siguiente tomará desde el 117 en adelante.

In [79]:
df.loc[117:]

sex agegrp    when   bp
patient                            
117      Female    60+  Before  161
117      Female    60+   After  152
118      Female    60+  Before  165
118      Female    60+   After  174
119      Female    60+  Before  149
119      Female    60+   After  151
120      Female    60+  Before  185
120      Female    60+   After  163

Si quisiéramos convertir de vuelta a patient en una variable:

In [80]:
df['Pacient'] = df.index
df

sex agegrp    when   bp  Pacient
patient                                     
1          Male  30-45  Before  143        1
1          Male  30-45   After  153        1
2          Male  30-45  Before  163        2
2          Male  30-45   After  170        2
3          Male  30-45  Before  153        3
...         ...    ...     ...  ...      ...
118      Female    60+   After  174      118
119      Female    60+  Before  149      119
119      Female    60+   After  151      119
120      Female    60+  Before  185      120
120      Female    60+   After  163      120

[240 rows x 5 columns]

Y si quiero volver los índices al default también lo puedo hacer usando `reset_index()` (en nuestro caso sería `df.reset_index()`

Puede interesarnos crear una nueva variable, a partir de una condición de otra variable.

In [ ]:

def calcular_riesgo(bp):
    if bp > 170:
        return 1
    else:
        return 0


df['riesgo'] = df['bp'].apply(calcular_riesgo)
print(df)

Quizás nos interese crear variables más sencillas

In [ ]:
#Si queremos darle el mismo valor a toda la serie
df["diabetic"]=False 
df

Puede que queramos eliminar una variable

In [ ]:
df = df.drop("diabetic", axis=1)
df

Axis=1 implica las columnas, axis=0 son las filas.

In [ ]:
# transformamos la columna a minusculas
df["when"] = df["when"].str.lower()
df

Puede que me interese asignar etiquetas a las variables:

Puede interesarnos quedarnos con una parte del dataframe, tomando una **muestra aleatoria**. Para esto usaremos `.sample(frac=0.2)` por ejemplo, para dar cuenta que queremos quedarnos con el 20% de la muestra. 

In [ ]:
df_sample=df.sample(frac=0.05)
df_sample.describe

# Valores missing

Como sabemos, muchas veces trabajamos con bases de datos que tienen valores missings. En primer lugar es importante poder saber que los enemos, para luego saber cómo trabajar con ellos. 

In [ ]:
dataset.info()

Notar que en la variable kilometraje tenemos 203 valores non-null, contra 258 de las otras columnas. Miremos más en detalle qué sucede utilizando la función `isna()` para identificar los datos ausentes: 

In [ ]:
dataset[dataset.Kilometraje.isna()]

La coincidencia es que todos los missings en km son los autos nuevos. Entonces, puede ser buena idea transformar esos valores NaN en 0. Lo haremos utilizando la función `fillna()`

Para [leer más](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html) sobre esta opción

In [ ]:
dataset.fillna(value = {'Kilometraje': 0}, inplace = True)
dataset

Notar que Kilometraje no era la única variable con problemas. Veamos qué sucede con la columna Valor.

In [ ]:
dataset[dataset.Valor.isna()]

In [ ]:
dataset["Valor"].describe()

Como vemos, tenemos valores -999999, lo que muchas veces se usa para especificar missings, y que puede ser buena idea modificar. Primero, convirtámoslos en missings.


In [ ]:
dataset["Valor"].replace(-999999, np.nan, inplace = True)
dataset["Valor"].describe()

Puede que ahora queramos eliminar las opciones con missings, eso lo haremos con la función `drop.na()`

In [ ]:
dataset.dropna(subset = ['Valor'], inplace = True)
dataset.info()

In [ ]:
dataset

# <CENTER> INTRODUCCIÓN A GRÁFICOS
    
 Una de las bibliotecas que usaremos para graficar es `matplotlib`. Debemos importarla.

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
dir(plt)

In [ ]:
dataset["Kilometraje"].hist()

Trabajaremos con gráficos más en profundidad la clase que viene.

# <CENTER> EJERCICIOS 

   ### Para resolver en grupo: 
    
   ### Elijan una de las siguientes tareas: 
    
    
- **Municipios 1:** 
    

    - Importar la base de municipios disponible en la carpeta de clase 2.
    
    - generar la variable de recaudación como la suma de las variables de las tasas (abl, tish, red vial) y multiplicada por un millón. 
    
    - generar la recaudación per cápita.
    

    
- **Municipios 2:** 
    
    
    - Importar la base de municipios disponible en la carpeta de clase 2.
    
    - Calcular cuántas personas tienen internet en el municipio.
    
    - Calcular la media de personas con internet según la actividad económica principal de los municipios. 
    
    
- **Bucles 1:**
    
    Crear un bucle donde el cajero de un comercio tenga que ingresar los valores de las distintas ventas que se hacen en el día. Los montos negativos no son válidos y deben avisarse al usuario. El bucle va acumulando los valores de las ventas del día. 
Al final del día, debe ingresar el valor 0 para cerrar la caja, y el programa devolverá el total recaudado.   
    
    
- **Bucles 2:** 
    
    Sean 2.3, 3.0, 2.5, 2.8, 2.1, 2.9, 3.2, 3.5, 2.7, 3.3 las tasas de crecimiento anuales del PBI de un país cuyo PBI inicial era de 1000 para los próximos 10 años. Calcule el PBI final y la tasa de crecimiento promedio.